packages/alfa-rules/src/sia-r62/rule.ts

```diff
@@ -230,21 +169,57 @@ function isDistinguishable(
   device: Device,
   context?: Context
 ): Predicate<Element> {
-  return (element) => {
-    if (
-      test(not(hasOutline(device, context)), container) &&
-      test(hasOutline(device, context), element)
-    ) {
-      return true;
-    }
+  return or(
+    // Things like text decoration and backgrounds risk blending with the
+    // container element. We therefore need to check if these can be distinguished
+    // from what the container element might itself set.
+    hasDistinguishableTextDecoration(container, device, context),
+    hasDistinguishableBackground(container, device, context),
+
+    // We consider the mere presence of borders or outlines on the element as
+    // distinguishable features. There's of course a risk of these blending with
+    // other features of the container element, such as its background, but this
+    // should hopefully not happen (too often) in practice. When it does, we
+    // risk false negatives.
+    hasOutline(device, context),
+    hasBorder(device, context)
+  );
+}
 
-    if (
-      test(not(hasTextDecoration(device, context)), container) &&
-      test(hasTextDecoration(device, context), element)
-    ) {
-      return true;
-    }
+function hasDistinguishableTextDecoration(
+  container: Element,
+  device: Device,
+  context?: Context
+): Predicate<Element> {
+  return (element) =>
+    test(not(hasTextDecoration(device, context)), container) &&
+    test(hasTextDecoration(device, context), element);
```

Comment:
We could have the reverse situation (container with text decoration and element without)
```suggestion
    test((hasTextDecoration(device, context)), container) !==
    test(hasTextDecoration(device, context), element);
```

Reply:
Text decoration is applied to all descendant text so the reverse situation won't result in the link being distinguishable:

```html
<!DOCTYPE html>

<style>
  p {
    text-decoration: underline;
  }

  a {
    text-decoration: none;
  }
</style>

<p>Hello <a href="#">world</a></p>
```

<img width="85" alt="rendered output" src="https://user-images.githubusercontent.com/1174718/114022938-947f5a80-9872-11eb-9869-db23c456d2ea.png">